### ISSUE TITLE
Export binary

### ISSUE THREAD

**Issue by necolas** [karma-runner/karma#1078]
The binary export was removed in this commit: c7d46270aca83ecfe78f69fa923bc574c0b5bfdc

But I think this has broken npm-scripts aliasing. Normally you can run the local install of a package from npm-scripts in package.json (because npm looks up the binary in `./node_modules/.bin/`:

```json
{
  "scripts": {
    "test": "karma start"
  }
}
```

But this doesn't work with Karma anymore.

**Comment by nzakas** [comment]
Agree with @necolas here. This is pretty confusing and frustrating. If you are shipping the binary in the `karma` package, why not include an easy way to use it?

**Comment by peter-mouland** [comment]
Hi,  just came across this as i was going to do the same PR.  In the docs it says you can use the full path.

https://karma-runner.github.io/0.12/intro/installation.html
 > # Run Karma:
 > $ ./node_modules/karma/bin/karma start

I dont get why i would want the cli when I would like to simply put `karma start xx.js` into my package.json.  Since the bin is in karma, and it works, could we add the bin into karma's package.json?

I think the route of asking people to also install the cli is making the process a bit convoluted.   The cli doesn't currently have tests and is only used for karma, so it isn't really a separate project. 

#1351 #1322 #1275 ... there are few issue open and closed about this!

just my 2cents.

**Comment by donaldpipowitch** [comment]
Sorry, I don't get it? Just made a pull request myself. That is quite annoying to realize that I wasted my time with this even if it was just some minutes. Installing a different module (karma-cli) is just unexpected and in my experience not idiomatic for a node module.

**Comment by nzakas** [comment]
@maksimr I'm trying to understand your objection, but I'm having a hard time. From our perspective, your stance is confusing because we are just asking for an easier way to run the binary that is already shipped in this package. As #1361 shows, it's basically a one-line change that would make our lives a lot easier. We aren't asking for any change in functionality, nor are we asking for additional files to be included. This is the logical equivalent of a symlink.

This package has 8x the downloads of `karma-cli`, which means the majority of users have no need for `karma-cli`. That, plus the number of issues and PRs related to this issue, seem to indicate a pain point for users. Is there anything we can do to change your mind?

**Comment by maksimr** [comment]
Guys thanks for your feedback.

@donaldpipowitch your PR once again says that we have a *problem* we need to solve.

@nzakas 
if we simple add link to binary in karma, it can lead to confusing behavior when we install karma globally. For example: you have karma version 0.12 in the project but globally you have karma version 0.11 and if you simple type `karma` in terminal then you run karma@0.11 instead karma@0.12. Yup, you can simple update you global karma to latest version, but what if you have more than one project locally which use different version of karma? In this case you can not have two instance of global version of karma. Yes, you can run test like `npm test` but now we depend from `npm` how cli launcher, in old version of npm you can not pass parameters(For example how you run `karma run --grep="..."` using npm).  
**(IMHO)**

For me the main problem of this changes is a habit, but we really have many requests `add karma to binary`. I think **we should discuss** this issue and fix it either add link on binary in karma or update documentation make it more clear.

Thanks

/cc @karma-runner/contributors @zzo @vojtajina

**Comment by nzakas** [comment]
I understand you are worried about `karma` possibly meaning two different things depending on where it is used. Would a solution be to expose the binary in this package with a different name? Maybe `karma-runner` instead of just `karma`?

**Comment by peter-mouland** [comment]
my opinion seems to ebb and flow the more i learn, but currently i ask users of my projects to always use npm commands, ie. `npm start`, `npm test`, `npm run build` etc.  This is because that way i make sure i never rely on global installs and i know exactly what version is being run.  This has the bonus of the casual dev not caring what technology those steps are using in the background and projects using different technologies still start in the same way.

Would it be acceptable to recommend people install karma locally, then add the `karma test ..` to the npm scripts object?

**Comment by nzakas** [comment]
@peter-mouland that's what people are doing already. What we are asking for is a shortcut to the executable. How people choose to wire up their build system to use it is a different story.

**Comment by peter-mouland** [comment]
thats good - i mentioned it because one reason given for having the karma-cli was due to global vs local installs.  If globals are discouraged then this reason is no longer a concern and the package.json `bin` string can be updated as suggested.

**Comment by nzakas** [comment]
Yeah, there's definitely a trend towards not doing global installs for CLIs because it makes projects no longer self-contained.

**Comment by tkrotoff** [comment]
@maksimr 
> For example: you have karma version 0.12 in the project but globally you have karma version 0.11 and if you simple type karma in terminal then you run karma@0.11 instead karma@0.12.

Very good example of why it is bad to install global npm packages.
IMHO this issue is out of scope: this is the way UNIX works (security reasons).

Example:
```Shell
cd ~
echo 'echo "Hello, World!"' > bin/ls2
chmod a+x bin/ls2
export PATH=$PATH:bin
ls2
# => Hello, World!
mv bin/ls2 bin/ls
ls
# => regular /bin/ls output, not ~/bin/ls output
./bin/ls
# => Hello, World!
```

Please, just add back `karma` to `node_modules/.bin`, kill `karma-cli` and stop bypassing the way the OS works (and do not advice for global packages otherwise first year students will have to learn how [PATH](https://en.wikipedia.org/wiki/PATH_(variable)#Unix_and_Unix-like) work and the `which` command).

If you really believe this is a critical "issue", then propose the npm team a fix for all npm packages.

**Comment by maksimr** [comment]
@tkrotoff Thanks

I like idea export binary with different name like `karma-bin` or `karma-runner` 
How [suggested](https://github.com/karma-runner/karma/issues/1078#issuecomment-86978471) @nzakas 

Then it should be clear if you use `karma-bin` then you use a specific version of karma.

**Comment by donaldpipowitch** [comment]
I'd like to see a solution here soon or else more pull requests will drop in. Here is one from 14 days ago: https://github.com/karma-runner/karma/pull/1384

**Comment by zzo** [comment]
My initial thought reading through all of this is to dump the global karma-cli completely and put the link in .bin/karma and be done.  A project has installed a specific version of karma they should use the 'karma' executable that comes with it.  I don't get why anything should be installed globally.  That seems to be the root of the problem.  Especially since the 'karma' executable is already being shipped with the 'karma' package.

**Comment by necolas** [comment]
Any chance the sensible, common thing can be done? Or shall I just close this year-long thread?

**Comment by zzo** [comment]
ha depends what you definition of 'sensible' is of course.
I'm having problems releasing a new version of karma - once that gets sorted out I will apply the patch linking 'karma' in node_modules/.bin and release it

**Comment by nzakas** [comment]
@necolas I'd suggest leaving it open. It seems @maksimr is open to this, we just need to gently nudge along. :)

**Comment by necolas** [comment]
Great, thanks

**Comment by jhabdas** [comment]
While you were sleeping, this issue was resolved with #1384.

**Comment by necolas** [comment]
Thanks
